### ISSUE TITLE
Support for Native ECS AutoScaling

### ISSUE THREAD

**Issue by snario** [convox/rack#1074]
I want to scale the `--count` of processes based on some CloudWatch alarms such as high average memory usage (high load, scale up) or low overall requests count (low load, scale down).

Right now I can't do that because the AutoScaling policies get deleted / a new service is created that replaces the old one.
1. Go to an ECS Service console page
2. Click Update
3. Set a policy and click save
4. Update Service
5. Service now has auto scaling rules
   …
6. Return a couple minutes later and the rules are gone

**Comment by andykent** [comment]
We would really love to get this functionality too, proper autoscaling of process counts would be a huge deal for us.

Are there any any plans to implement this in the near term? 
Or does anyone have any suggested solutions that might get us part of the way there?

**Comment by andykent** [comment]
To Clarify our use-case.

We have our rack set to autoscale but if we get spikes of traffic though-out the day it would be great to be able to scale a process based on traffic volume, currently this needs manual intervention. 

From the ECC docs it looks like you can now configure ECS autoscaling rules to increase the process count based on CPU or Mem utilization. We'd love it to be latency based but it doesn't look like that is possible currently.

I'm thinking having rules of Min/Max/Desired count and CPU/Mem alarm thresholds included in the docker-compose.yml would be the ideal solution for us. Maybe something like the following could work...

```
- convox.autoscaling.minimum=2
- convox.autoscaling.desired=2
- convox.autoscaling.maximum=10
- convox.autoscaling.cpu=75%
- convox.autoscaling.memory=90%
```

**Comment by andykent** [comment]
I just discovered that you can actually setup ECS autoscaling rules based on any cloudwatch alarms so, revising my above comment, I'd love to also support ELB latency based scaling. E.g. 

- scale up when average latency > 500ms for more than 2 consecutive 1 min windows. 

Given there are so many options/params for CW alarms around aggregation, threshold, window, occurrences, I'm starting to think it would be a nightmare to support all use-cases.

My simplified and more flexible proposal would be to allow end users to configure CW alarms in the Web UI however they like and then pass them in via docker compose. So maybe something like...

```
- convox.autoscaling.minimum=2
- convox.autoscaling.maximum=10
- convox.autoscaling.add.alarm=my-high-latency-alarm
- convox.autoscaling.add.cooldown=300
- convox.autoscaling.remove.alarm=my-normal-latency-alarm
- convox.autoscaling.remove.cooldown=300
```

**Comment by bejmuller** [comment]
+1

**Comment by beedub** [comment]
AWS autoscaling policies should persist as long as the rack includes the change from custom ecs service to native ecs service (https://github.com/convox/rack/pull/1081).

The problem now is that the native ecs service requires and enforces the desired count on each cloudformation update, so each service's desired count is reset to the value specified in the template instead of keeping the live value (which autoscale may have adjusted). (To be clear, this likely would have been a problem with the old custom ecs service as-it-was but it would have allowed rack to customize the behavior to set the desired count to the live count on updates)

The current workaround we are attempting is to sync the service's desired counts to the template before deploying and scaling but it adds another piece of complexity and failure.

**Comment by mattmanning** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Application/Container Autoscaling](https://trello.com/c/dNQt0uTl/1145-application-container-autoscaling)

**Comment by idris** [comment]
I think this belongs in app params, or some other CLI/console interface, rather than as labels in docker-compose.

It seems like scaling, capacity, etc. is handled outside of code in Convox.  That makes sense, since you wouldn't want to deploy any time you want to change these settings.

**Comment by soulshake** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [ECS autoscaling](https://trello.com/c/ZYLqsaH5/52-ecs-autoscaling)

**Comment by perryn** [comment]
Hi @beedub: Did you end up implementing the workaround you described? How did it go?

**Comment by beedub** [comment]
@perryn we ended up with something similar but not exactly the same.

We used terraform to create autoscaling on convox-created ecs services, and wrote a script that runs every minute to update CF stack with the current count, which fluctuates based on the AWS app autoscaling.

The only annoyance is that only one stack update can take place at a time, so other stack updates (like deployment) need to be queued, but otherwise, it's working acceptably.
